_posts/2025-02-13-rfprs.md

```diff
@@ -0,0 +1,372 @@
+let you know of things that you may not have considered, or why this particular algorithm you've
+employed may have negative effects on the backend APIs it eventually calls. Style nits are OK but
+avoiding an incident by pointing out an unhandled edge case is much better.
+
+The reviewers only have so much time to devote to your PR. If that time is largely spent being
+mystified about the changeset, that is time that they are not spending giving you the quality
+feedback you should be hoping for. The time a reviewer spends on any particular PR is then
+multiplied by the number of people that end up reviewing it.
+
+> Time spent creating a better PR costs the business less to merge it.
+
+# RFPR Overview
+
+Reviewer First Pull Requests (RFPRs) are a prescriptive approach to building better PRs that are
+cheaper to review. At Mesosphere I worked on a team under [Kevin
+Klues](https://www.linkedin.com/in/klueska). After creating a PR which was hard to review he
+convinced me that his "Reviewer-First" methodology was worth adopting, and I've been promoting it
+ever since.
+
+> At the heart of RFPRs is the concept that PRs, like a book, should tell a story.
+
+When reading a book, you typically know what it's about before you start because you've read the
+intro on the back. You start at the first chapter, and continue reading until the end. Even if the
+book was particularly long or complex, the chapters were structured in a way to guide you through
+the story that the author wanted to tell.
+
+RFPRs operate under the same principle. A reviewer should be able to open the PR and read the
+description to get a good idea for what the PR is addressing and what to expect inside. RFPRs are
+structured so that they can be reviewed commit-by-commit, in order, leaving feedback as they go.
+
+> Large or complex PRs are made more approachable if the author guides the reviewers.
+
+# RFPR Guidelines
+
+Following are guidelines for creating RFPRs. Afterwards we'll discuss some tooling that will make
+following the guidelines much easier.
+
+## PR titles
+
+PR titles should describe what the PR accomplishes.
+
+Examples of good PR titles:
+
+- [CONTROL_42] Expose gRPC endpoints for all controlplane APIs.
+- kafka: Create new topic for user events and enable mirrormaker for all regions.
+- Modify CMS domain models to use composition instead of inheritance.
+
+PR titles which should be avoided:
+
+- [CONTROL_42] (just a JIRA ticket number)
+- fix bug
+- (nothing)
+
+Imagine that you are trying to find a particular PR on GitHub during an incident. The PR title
+should be descriptive enough so that it can be easily identified.
+
+## PR descriptions
+
+The PR description sets the stage for the reviewer. Since they are going to review the PR
+commit-by-commit, the description should give them as much context as possible before they start.
+Assume that the reviewer does not have the context that you do.
```

Comment:
I have to admit, I honestly don't usually review PRs commit-by-commit, for many of the reasons you document here -- people haven't practiced this discipline. I wonder how many people actually know how to review PRs commit-by-commit in Github and other tools. It could be worth demonstrating how to do this since most folks likely just pull up the entire changeset at once.

Reply:
That's a great callout. I'll add an appendix for this.

Reply:
I explicitly call out at the top of the PR that suggest reviewers review commit-by-commit. I'm a broken record at this point. I think the default mode is to just slog through the full diff.